### PR TITLE
Fix database path for user provided YAML file

### PR DESCRIPTION
**Pull request by wigging**
## Fixes/Resolves:

Fixes issue #907.

## Summary/Motivation:

The Database class has an optional argument where the user can provide the path to a YAML file for use with a zero-order model.

```python
from pyomo.environ import ConcreteModel
from watertap.core.wt_database import Database

model = ConcreteModel()
model.db = Database(dbpath="./myfile.yaml")
```

This path variable does not account for the name of the YAML file; therefore, the default database file in `watertap/data/techno_economic`/ is always used for the associated model. This pull request fixes the Database class to actually use the YAML file at the given path.

## Changes proposed in this PR:
- Use a `self._filename` attribute for the `Database` class to define the `technology` variable. This will be set when `dbpath` is given; otherwise, the default database file is used for the associated zero-order model.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/wt_database.py [hunk 1]
```diff
@@ -44,7 +47,8 @@ def __init__(self, dbpath=None):
                 "techno_economic",
             )
         else:
-            self._dbpath = dbpath
+            self._dbpath = os.path.dirname(dbpath)
+            self._filename = os.path.basename(dbpath).split(".")[0]
```
Comment (lbianchi-lbl) [inline watertap/core/wt_database.py]: Suggestion using `pathlib.Path`:

```suggestion
            dbpath = Path(dbpath).resolve()
            self._dbpath = dbpath.parent
            self._filename = dbpath.stem
```

File: watertap/core/wt_database.py [hunk 2]
```diff
@@ -196,7 +200,10 @@ def _get_technology(self, technology):
             # If data is already in cached files, return
             return self._cached_files[technology]
         else:
-            # Else load data from required file
+            # If user provided a YAML file, use the filename as the technology
+            if hasattr(self, "_filename"):
+                technology = self._filename
+            # Load from database YAML file or from user provided YAML file
             try:
                 with open(os.path.join(self._dbpath, technology + ".yaml"), "r") as f:
```
Comment (lbianchi-lbl) [inline watertap/core/wt_database.py]: Suggestion: the contents of the `try` block can be simplified by e.g.:

```py
technology_file_path = (self._dbpath / technology).with_suffix(".yaml")
lines = technology_file_path.read_text()
```

File: watertap/core/tests/test_wt_database.py [hunk 3]
```diff
@@ -54,7 +53,7 @@ def test_custom_path():
     )
 
     assert db._dbpath == os.path.join(
-        os.path.dirname(os.path.abspath(__file__)), "..", "..", "core"
+        os.path.dirname(os.path.abspath(__file__)), "..", ".."
```
Comment (lbianchi-lbl) [inline watertap/core/tests/test_wt_database.py]: Clarification: if the purpose of this test is just to check that `db._dbpath` is processed correctly using any arbitrary path, why not to use `Path(__file__).resolve()` (i.e. the absolute path to the test file itself)? Barring very exotic corner cases, this would be guaranteed to exist almost by definition.

(Also, feel free to disregard if this implementation detail predates this PR and is unrelated to the other changes).

File: watertap/core/tests/test_zero_order_sido_reactive.py [hunk 4]
```diff
@@ -48,7 +48,7 @@
 
 solver = get_solver()
 
-local_path = os.path.dirname(os.path.abspath(__file__))
+local_path = os.path.dirname(os.path.abspath(__file__)) + "/test_sidor_data.yaml"
```
Comment (lbianchi-lbl) [inline watertap/core/tests/test_zero_order_sido_reactive.py:51]: Suggestion: using `pathlib.Path`:

```suggestion
local_path = Path(__file__).resolve().parent / "test_sidor_data.yaml"
```

File: watertap/core/wt_database.py [hunk 5]
```diff
@@ -196,7 +200,10 @@ def _get_technology(self, technology):
             # If data is already in cached files, return
             return self._cached_files[technology]
         else:
```
Comment (lbianchi-lbl) [inline watertap/core/wt_database.py]: Suggestion: the `if` block above contains a `return` statement, so the `else` is not necessary.

**Review by lbianchi-lbl — COMMENTED** [review]
I've left a few comments with suggestion on the implementation, but nothing show-stopping.

File: watertap/core/wt_database.py (hunk 1 above)
Comment (wigging) [reply watertap/core/wt_database.py]: I did not use `pathlib` because `os` was already used in this file. Is there any reason why `os` was used instead of `pathlib` for this class?

Comment (wigging) [reply watertap/core/wt_database.py]: Are there other areas of watertap that would expect an `os` path instead of a `pathlib` path?

File: watertap/core/wt_database.py (hunk 2 above)
Comment (wigging) [reply watertap/core/wt_database.py]: This assumes using `pathlib` instead of `os`. See my previous comment about using `pathlib` instead of `os` for paths.

File: watertap/core/tests/test_wt_database.py (hunk 3 above)
Comment (wigging) [reply watertap/core/tests/test_wt_database.py]: All of the original code in this pull request predates me. So I don't know why it's pointing to `core`.

File: watertap/core/tests/test_zero_order_sido_reactive.py (hunk 4 above)
Comment (wigging) [reply watertap/core/tests/test_zero_order_sido_reactive.py:51]: See previous comment about using `pathlib` vs `os`.

File: watertap/core/wt_database.py (hunk 5 above)
Comment (wigging) [reply watertap/core/wt_database.py]: This whole function is badly written. I just made the changes needed to get the path functionality working.

File: watertap/core/wt_database.py (hunk 1 above)
Comment (bknueven) [reply watertap/core/wt_database.py]: > I did not use `pathlib` because `os` was already used in this file. Is there any reason why `os` was used instead of `pathlib` for this class?

I don't think so. We should switch to pathlib where/when convenient.

File: watertap/core/wt_database.py (hunk 5 above)
Comment (wigging) [reply watertap/core/wt_database.py]: Are there any examples of `self._cached_files` being used?

File: watertap/core/wt_database.py (hunk 1 above)
Comment (wigging) [reply watertap/core/wt_database.py]: Ok, I replaced os with pathlib.

File: watertap/core/wt_database.py (hunk 2 above)
Comment (wigging) [reply watertap/core/wt_database.py]: Done.

File: watertap/core/tests/test_wt_database.py (hunk 3 above)
Comment (wigging) [reply watertap/core/tests/test_wt_database.py]: Implemented this with pathlib instead of os.

File: watertap/core/tests/test_zero_order_sido_reactive.py (hunk 4 above)
Comment (wigging) [reply watertap/core/tests/test_zero_order_sido_reactive.py:51]: Done.

File: watertap/core/tests/test_wt_database.py [hunk 6]
```diff
@@ -14,48 +14,28 @@
...
-    )
+    assert db._dbpath == Path(__file__).parent.parent.parent / "data/techno_economic"
 
 
 @pytest.mark.unit
 def test_invalid_path():
-    with pytest.raises(
-        OSError,
-        match="Could not find requested path foo. Please "
-        "check that this path exists.",
```
Comment (bknueven) [inline watertap/core/tests/test_wt_database.py:42]: We generally like having these match statements so we can be sure the "right" error was raised. But perhaps not strictly needed.

**Review by bknueven — APPROVED** [review]

File: watertap/core/tests/test_wt_database.py (hunk 6 above)
Comment (wigging) [reply watertap/core/tests/test_wt_database.py:42]: Pushed up a commit to address this.

File: watertap/core/wt_database.py (hunk 5 above)
Comment (wigging) [reply watertap/core/wt_database.py]: I removed the `else` statement but left the `self._cached_files` since I'm not sure if these are used elsewhere in the code.

**Review by bknueven — COMMENTED** [review]
Something worth mentioning is I think this breaks the original intent of supplying your own `dbpath`: that you have a whole set of `yaml` files just like in `data/techno_economic/`. Instead you have to supply individual file names.

Maybe this makes more sense for our workflow, where maybe people will come with their own custom units. But in that case shouldn't they just use a separate "Database" object with the custom units?

**Comment by wigging** [comment]
@bknueven I don't know the original intentions for this database class. But based on the tests (there are no examples that I'm aware of), the `dbpath` is being used to parse a single YAML file. I don't see any tests or examples that demonstrate using the path to parse several YAML files in a directory.

**Comment by wigging** [comment]
Looks like @andrewlee94 is the creator of the `Database` class. Andrew, what is the purpose of the `dbpath` argument? From looking at the tests, it is used to parse a single YAML file. I'm not aware of any examples where it is used to parse many YAML files at once in a directory. Or maybe I missed something, I don't know. Any clarification on this would be helpful.

**Comment by andrewlee94** [comment]
It has been a very long time since I wrote this, but I believe it was intended to be a path to a folder containing a set of YAML files.

**Comment by bknueven** [comment]
> @bknueven I don't know the original intentions for this database class. But based on the tests (there are no examples that I'm aware of), the `dbpath` is being used to parse a single YAML file. I don't see any tests or examples that demonstrate using the path to parse several YAML files in a directory.

The default `dbpath` is this whole directory: https://github.com/watertap-org/watertap/tree/main/watertap/data/techno_economic. This test points to path, not a filename: https://github.com/watertap-org/watertap/blob/85e1f521141dd52c10c560d52b651e22c4832311/watertap/core/tests/test_wt_database.py#L48-L58

Here is an example flowsheet: https://github.com/watertap-org/watertap/blob/main/watertap/examples/flowsheets/case_studies/municipal_treatment/municipal_treatment.py. Note only one `Database` object, lots of unit models.

 I'm going to revisit #907.

**Review by bknueven — CHANGES_REQUESTED** [review]
After thinking more carefully about #907, this code works as intended.

I think there's a separate issue of *how* we associate filenames with unit models, but I don't think requiring a separate database for each unit model is the way to go. This will rapidly become annoying if a user has an entire set of yaml files.

It might be that better documentation is needed.

**Comment by wigging** [comment]
Would something like the following suffice:

```python
class Database:
    def __init__(self, dbpath=None, yamlfile=None):
        # blah blah blah
```

The user could provide the path to their YAML file like this:

```python
model = ConcreteModel()
model.db = Database(yamlfile="./myparams/dual_media_filter.yaml")
```

Actually, is there any scenario when `dbpath` would point any where other than `watertap/data/techno_economic/`?

**Comment by bknueven** [comment]
> The user could provide the path to their YAML file like this:
> ```python
> model = ConcreteModel()
> model.db = Database(yamlfile="./myparams/dual_media_filter.yaml")
> ```

What I don't like about this is it turns the `Database` into a manager for a specific yaml file as opposed to a manager for a directory of yaml files. There's also the issue with `water_sources`.


I think the thing that makes more sense is for the *unit model* to tell the database where its yaml file is, e.g.,
```python
model = ConcreteModel()
model.db = Database(dbpath=".")

model.fs = FlowsheetBlock(dynamic=False)
model.fs.params = WaterParameterBlock(solute_list=["nonvolatile_toc", "toc", "tss"])

model.fs.unit = DualMediaFiltrationZO(property_package=model.fs.params, database=model.db, yaml_file="./myparams/dual_media_filter.yaml")
```

This would allow the user to bind an instance of a ZO unit model class to their own YAML file, which seems like a good use-case: the user can customize a single unit in the process easily while taking the defaults for all the other units.

**Comment by wigging** [comment]
So what's the point of having the user define `dbpath` for the `Database`? Why would a user ever want to modify that path? 

If the user did something like shown below then why define a path in two different places?

```python
model = ConcreteModel()
model.db = Database(dbpath=".")

model.fs = FlowsheetBlock(dynamic=False)
model.fs.params = WaterParameterBlock(solute_list=["nonvolatile_toc", "toc", "tss"])

model.fs.unit = DualMediaFiltrationZO(property_package=model.fs.params, database=model.db, yaml_file="./myparams/dual_media_filter.yaml")
```

If a user provides the YAML file, then why does a database instance need to be created? Can this be done:

```python
model = ConcreteModel()

model.fs = FlowsheetBlock(dynamic=False)
model.fs.params = WaterParameterBlock(solute_list=["nonvolatile_toc", "toc", "tss"])

model.fs.unit = DualMediaFiltrationZO(property_package=model.fs.params, yaml_file="./myparams/dual_media_filter.yaml")
```

**Comment by bknueven** [comment]
> So what's the point of having the user define `dbpath` for the `Database`? Why would a user ever want to modify that path? 


Copy the contents of `watertap/data/techno_economic` to your own directory, make the modifications you like, and point the `dbpath` to the new directory. We could probably define a simple function to "set up" the database files in a desired location.

**Comment by wigging** [comment]
I'm so confused. Why would I copy the entire database directory of YAML files to some other location? That directory is part of watertap.

**Comment by bknueven** [comment]
So you can easily set the database parameters to your own values.

**Comment by wigging** [comment]
I installed watertap in editable mode with `pip install -e`. I don't need to copy any of the database files. I can edit them directly. Anyway, that is not the point of this pull request.

**Comment by andrewlee94** [comment]
@wigging The point is that there is no single set of universal parameters for this - different projects/applications might have different sets of parameters and a single user might need to work on multiple projects. Thus, there needs to be a way for users to select from different parameter sets.

**Comment by wigging** [comment]
@andrewlee94 Why use a YAML file in `watertap/data/techno_economic` to define the parameter values for these models? Why not set those values in the model itself?

**Comment by andrewlee94** [comment]
@wigging There is nothing to prevent the user from doing that, but the whole zero-order model thing grew out of WaterTAP3 where it was intended to be a very hands-off experience for low-level users. Plus, these parameters tend to be case specific, but there are often many unit instances in a single case - the goal was to have a central location for defining the global parameters, but have a way that they could be easily swapped out for a different set (note that we generally expected the person creating the datasets to have a greater knowledge of how things worked).

**Comment by wigging** [comment]
@andrewlee94 Can you point me to the documentation on how to setup and run a zero-order model? I've been looking at the tests inside of watertap as a guide. There's a section on [zero-order unit models](https://watertap.readthedocs.io/en/stable/technical_reference/unit_models/zero_order_unit_models/index.html) in the docs but it just describes each model, there's nothing about actually setting up the model and running it.

**Comment by andrewlee94** [comment]
This is the only example I could find in a quick search: https://github.com/watertap-org/watertap/blob/main/watertap/examples/flowsheets/case_studies/municipal_treatment/municipal_treatment.py

However, I am not the one to ask about that - I was responsible for putting together the basic infrastructure but not putting it to use.

**Comment by wigging** [comment]
@andrewlee94 Yeah, I saw these scripts which demonstrate using zero-order models within a flowsheet environment. But I was wondering if there is documentation regarding running an individual zero-oder model. So apparently there is no documentation. Just tests and scripts buried inside of the watertap source code.

**Comment by adam-a-a** [comment]
> @andrewlee94 Yeah, I saw these scripts which demonstrate using zero-order models within a flowsheet environment. But I was wondering if there is documentation regarding running an individual zero-oder model. So apparently there is no documentation. Just tests and scripts buried inside of the watertap source code.

The only other documentation we have on zero-order modeling is [here](https://watertap.readthedocs.io/en/stable/technical_reference/core/index.html). As far as putting documentation together on the workflow of putting all the pieces together for creating and running zero-order models, that's still a subtask in a master issue we have that relates to this: #595. Notably, we haven't had this as high-priority since we were in the weeds for quite some time with other issues of importance. So yes, we typically follow test files as examples to go by, and for developing and running individual unit models, we'd usually just meet and go over it with each other (or with someone who was interested in using ZO models). We still need someone to step up and get some documentation in on ZO modeling development/usage workflows. Interested? =)

**Comment by adam-a-a** [comment]
@wigging  So I thought about this more, and it seems to me that there are a couple different aspects we need to consider here. You should be able to navigate to some other db path aside from `data/technoeconomic`. Fair enough. I can see the case where someone might want to have their own directory of yaml files to work from. 

However, for your purposes, and please correct me if I'm wrong, it seemed like you wanted to use a unit model that already existed and to supply your own yaml file to populate that unit model. From the developer perspective, the way I would suggest doing this is by adding a subtype to the existing yaml attributed to the unit model you're working with. That is, add to the existing yaml file, then specify your subtype name when using the unit model (if you're unsure how to do this due to lack of documentation, happy to have a call and talk through). On the other hand, if you are adamant about the idea that a user shouldn't have to dig into our yaml files and modify anything (seems reasonable that we shouldn't force users to do that), then I do think it makes sense for the custom yaml file to be specified at the unit level, overriding usage of the Database instance for that particular unit.

As @andrewlee94 mentioned, this _all_ sprouted from an abrupt need to review nearly 100 simple models from another repo, cleanup and standardize their implementation, and migrate them into WaterTAP, all in a fairly short time. This might be why we didn't thoroughly consider other use cases that might come about. Once we accomplished the migration of those models, we were content with moving on to more important tasks, like working on detailed, predictive models which can be of far greater use. All that said, it is fairly easy to use the ZO modeling "framework," once you know how things connect. So, if the concern is implementing your use case, I would think we could do that easily (i.e., simulating a unit or flowsheet with your own specified parameter values). If the main concern is having a workflow for a novice user (not developer) to be able to do their own thing outside of the existing "database" and workflow that we've used nominally in the past (i.e., change the workflow so that the user can have more options/control over how they provide YAML data), then I wonder if the end justifies the means. How far do we take this?

File: watertap/core/wt_database.py [hunk 7]
```diff
@@ -192,23 +190,25 @@ def _return_component_list(self):
...
-                with open(os.path.join(self._dbpath, technology + ".yaml"), "r") as f:
-                    lines = f.read()
-                    f.close()
-            except OSError:
-                raise KeyError(f"Could not find entry for {technology} in database.")
 
-            fdata = yaml.load(lines, yaml.Loader)
+        # If user provided a YAML file, use the filename as the technology
+        if hasattr(self, "_filename"):
+            technology = self._filename
```
Comment (adam-a-a) [inline watertap/core/wt_database.py:199]: Shouldn't we have a test that includes a Database instance along with a unit model that makes use of this user-provided file path? I would want to see if anything breaks when using the "typical" sequence of methods (e.g., `load_parameters_from_database`, `get_unit_operation_parameters`) that use in our tests for ZO unit models and/or flowsheets.

**Comment by wigging** [comment]
I'm going to close this based on the discussion during Wednesday's meeting. Instead of restructuring the watertap code, I'll submit a pull request to improve the zero-order model documentation to make it clear how to properly use a zero-order model. We can discuss options for improving the zero-order models at a later date.
